cardano-testnet/src/Testnet/Start/Types.hs

```diff
@@ -104,16 +111,49 @@ instance Default CreateEnvUpdateTime where
   def = CreateEnv
 
 data CreateEnvOptions = CreateEnvOptions
-  { ceoTopologyType :: TopologyType
+  { ceoOnChainParams :: TestnetOnChainParams
+  , ceoTopologyType :: TopologyType
   , ceoUpdateTime :: CreateEnvUpdateTime
   } deriving (Eq, Show)
 
 instance Default CreateEnvOptions where
   def = CreateEnvOptions
-    { ceoTopologyType = def
+    { ceoOnChainParams = def
+    , ceoTopologyType = def
     , ceoUpdateTime = def
     }
 
+data TestnetOnChainParams
+  = DefaultParams
+  -- | A file path to a JSON file containing on-chain params, formatted as:
+  -- https://docs.blockfrost.io/#tag/cardano--epochs/GET/epochs/latest
+  | OnChainParamsFile FilePath
+  | OnChainParamsMainnet
+  deriving (Eq, Show)
+
+instance Default TestnetOnChainParams where
+  def = DefaultParams
+
+data UserProvidedGeneses = UserProvidedGeneses
+  { upgShelleyGenesis :: UserProvidedData ShelleyGenesis
+  , upgAlonzoGenesis :: UserProvidedData AlonzoGenesis
+  , upgConwayGenesis :: UserProvidedData ConwayGenesis
+  } deriving (Eq, Show)
+
+instance Default UserProvidedGeneses where
+  def = UserProvidedGeneses
+    def
+    def
+    def
+
+-- | An HTTP request to get a file containing up-to-date mainnet on-chain parameters.
+-- The file should be formatted with Blockfrost format:
+-- https://docs.blockfrost.io/#tag/cardano--epochs/GET/epochs/latest/parameters
+mainnetParamsRequest :: HTTP.Request
+mainnetParamsRequest = unsafePerformIO $ HTTP.parseRequest
```

Comment:
That's just `MonadThrow`, so you can just do:
```suggestion
mainnetParamsRequest = fromJust $ HTTP.parseRequest
```
if you're certain that the url parsing works.